File: frontend/src/app/models/GoodsReceiving.ts
import { Supplier } from '../components/suppliers/supplier/supplier';
import { BaseModel } from './BaseModel';
import { CompanyDto } from './Company';
import { PurchaseOrderDto, PurchaseOrderItem } from './PurchaseOrder';
import { UserDto } from './User';

export interface GoodsReceivingDto extends BaseModel {
  grnNo: string;
  purchaseOrderId: string;
  purchaseOrder: PurchaseOrderDto;
  supplierId?: string;
  supplier?: Supplier;
  companyId?: string;
  company?: CompanyDto;
  receivedDate: Date;
  supplierDONo: string;
  supplierDODate: Date;
  supplierDOAttachment: string;
  status: string;
  remarks: string;
  gross: number;
  discount: number;
  totalAmount: number;
  createdById: string;
  createdBy: UserDto;
  isPostedToInventory: boolean;
  goodsReceivingItems: GoodsReceivingItem[];
}

export interface GoodsReceivingItem extends BaseModel {
  goodsReceivingId: string;
  goodsReceiving: GoodsReceivingDto;
  purchaseOrderItemId: string;
  purchaseOrderItem: PurchaseOrderItem;
  receivedQuantity: number;
  unitPrice: number;
  unit: string;
  totalPrice: number;
  remarks: string;
  invoiceQuantity: number;
}

export interface GoodsReceivingItemRequest {
  purchaseOrderItemId: string;
  receivedQuantity: number;
  unit: string;
  unitPrice: number;
  discount: number;
  totalPrice: number;
  remarks: string;
}

export interface CreateGoodsReceivingRequest {
  grnNo: string;
  purchaseOrderId: string;
  supplierId: string;
  companyId?: string;

  receivedDate: Date;
  supplierDONo: string;
  supplierDODate: Date;
  supplierDOAttachment: string;
  remarks: string;
  gross: number;
  discount: number;
  totalAmount: number;
  goodsReceivingItems: GoodsReceivingItemRequest[];
}

export interface UpdateGoodsReceivingItemRequest {
  id: string;
  purchaseOrderItemId: string;
  receivedQuantity: number;
  unit: string;
  unitPrice: number;
  discount: number;
  totalPrice: number;
  remarks: string;
}

export interface UpdateGoodsReceivingRequest {
  id: string;
  grnNo: string;
  purchaseOrderId: string;
  supplierId: string;
  companyId?: string;

  receivedDate: Date;
  supplierDONo: string;
  supplierDODate: Date;
  supplierDOAttachment: string;
  remarks: string;
  gross: number;
  discount: number;
  totalAmount: number;
  goodsReceivingItems: UpdateGoodsReceivingItemRequest[];
}

export interface GRNDropdownDto {
  purchaseOrders: PurchaseOrderDto[];
  suppliers: CompanyDto[];
  companies: CompanyDto[];
}

export interface CreateInvoiceFromGRNItemsRequest {
  supplierId: string;
  items: CreateInvoiceItemFromGRN[];
}

export interface CreateInvoiceItemFromGRN {
  goodsReceivingItemId: string;
  quantity: number;
}

File: backend/Model/GoodsReceiving.cs
namespace YLWorks.Model
{
    public class GoodsReceiving : BaseEntity
    {
        public Guid Id { get; set; }
        public string GRNNo { get; set; } = string.Empty;
        public Guid PurchaseOrderId { get; set; }
        public PurchaseOrder PurchaseOrder { get; set; } = null!;
        public Guid SupplierId { get; set; }
        public Company Supplier { get; set; }
        public DateTime ReceivedDate { get; set; }

        public string? SupplierDONo { get; set; }
        public DateTime? SupplierDODate { get; set; }
        public string? SupplierDOAttachment { get; set; }

        public string Status { get; set; } = "Draft"; //Draft, Partial, Completed
        public string? Remarks { get; set; }
        public decimal? Gross { get; set; }
        public decimal? Discount { get; set; }
        public decimal? TotalAmount { get; set; }
        public Guid? CreatedById { get; set; }
        public User? CreatedBy { get; set; }
        public ICollection<GoodsReceivingItem> GoodsReceivingItems { get; set; } = new List<GoodsReceivingItem>();
    }

    public class GoodsReceivingItem : BaseEntity
    {
        public Guid Id { get; set; }
        public Guid GoodsReceivingId { get; set; }
        public GoodsReceiving GoodsReceiving { get; set; } = null!;
        public Guid? PurchaseOrderItemId { get; set; }
        public PurchaseOrderItem? PurchaseOrderItem { get; set; } = null!;

        public decimal ReceivedQuantity { get; set; }
        public decimal? UnitPrice { get; set; }
        public decimal? Discount { get; set; }
        public string? Unit { get; set; }
        public decimal? TotalPrice { get; set; }
        public string? Remarks { get; set; }
    }

    public class GoodsReceivingItemRequest
    {
        public Guid? PurchaseOrderItemId { get; set; }
        public decimal ReceivedQuantity { get; set; }
        public decimal? UnitPrice { get; set; }
        public decimal? Discount { get; set; }
        public string? Unit { get; set; }
        public decimal? TotalPrice { get; set; }
        public string? Remarks { get; set; }
    }

    public class CreateGoodsReceivingRequest
    {
        public string GRNNo { get; set; } = string.Empty;

        public Guid PurchaseOrderId { get; set; }
        public Guid SupplierId { get; set; }

        public DateTime ReceivedDate { get; set; }

        public string? SupplierDONo { get; set; }
        public DateTime? SupplierDODate { get; set; }
        public IFormFile? SupplierDOAttachment { get; set; }

        public string? Remarks { get; set; }
        public decimal? Gross { get; set; }
        public decimal? Discount { get; set; }
        public decimal? TotalAmount { get; set; }

        public List<GoodsReceivingItemRequest> GoodsReceivingItems { get; set; } = new();
    }

    public class UpdateGoodsReceivingItemRequest
    {
        public Guid? Id { get; set; }

        public Guid? PurchaseOrderItemId { get; set; }
        public decimal ReceivedQuantity { get; set; }
        public decimal? UnitPrice { get; set; }
        public decimal? Discount { get; set; }
        public string? Unit { get; set; }
        public decimal? TotalPrice { get; set; }
        public string? Remarks { get; set; }
    }

    public class UpdateGoodsReceivingRequest
    {
        public Guid Id { get; set; }

        public string GRNNo { get; set; } = string.Empty;

        public Guid PurchaseOrderId { get; set; }
        public Guid SupplierId { get; set; }

        public DateTime ReceivedDate { get; set; }

        public string? SupplierDONo { get; set; }
        public DateTime? SupplierDODate { get; set; }
        public IFormFile? SupplierDOAttachment { get; set; }

        public string? Remarks { get; set; }

        public decimal? Gross { get; set; }
        public decimal? Discount { get; set; }
        public decimal? TotalAmount { get; set; }

        public List<UpdateGoodsReceivingItemRequest> GoodsReceivingItems { get; set; } = new();
    }

    public class GRNDropdownDto
    {
        public List<PurchaseOrder> PurchaseOrders { get; set; } = new();
        public List<Company> Suppliers { get; set; } = new();
        public List<Company> Companies { get; set; } = new();

    }
}